### PR TITLE
20% more performance for popFront(char[])

### PR DESCRIPTION
**Pull request by UplinkCoder**
As per title.
On average 15% more performance because of smaller code and avoiding bsr

### CODE REVIEW

**Comment by JackStouffer** [comment]
Can you post some measurements?

**Comment by UplinkCoder** [comment]
okay some measurements
```
pfold took 122us
pfnew took 116us
pfold took 122us
pfnew took 116us
pfold took 124us
```

**Comment by wilzbach** [comment]
> Can you post some measurements with code?

Maybe we can use this incident to start benchmarking directly in Phobos. A fancy way would be to run benchmarks for every PR to prevent performance regression, but a simpler approach is to at least collect them at a common place (e.g. `benchmarks`), s.t. (1) in the future enabling automated testing is possible and (2) other people can use these benchmarks for improving upon them (or manually checking regressions).

**Comment by JackStouffer** [comment]
> okay some measurements

This doesn't tell me anything. What's the code? What's the data you're using? Are you testing just Unicode strings or are you testing with mixed ascii and Unicode? Etc.

**Comment by UplinkCoder** [comment]
@JackStouffer I am using ASCII mixed with unicode, 80% Ascii, 20% unicode.
The file is called blns.txt
https://github.com/minimaxir/big-list-of-naughty-strings/blob/master/blns.txt
The code of the benchmark is here
http://paste.ofcode.org/3CpieAhkrTYEcSncbPKbrj

**Comment by UplinkCoder** [comment]
I am wondering about the freebsd failure. 
It seems related but I have no idea why it occurs.

File: std/range/primitives.d [hunk 1]
```diff
@@ -2104,28 +2104,28 @@ if (isNarrowString!(C[]))
...
+            2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+            2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+            3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
+            4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 6, 6, 1, 1
+        ];
+        /*
+          This implementation of popFront does not special case 0b10xx_xxxx
+          Nither did the one before.
+          on illigal strings we may skip too many bytes.
+         */
```
Comment (andralex) [inline std/range/primitives.d]: I don't understand this comment. So if the first byte has the shape `0x10xx_xxxx` (illegal) we skip one byte which is what we need to do.

**Comment by andralex** [comment]
Seems the autotester failure is a timeout.

File: std/range/primitives.d (hunk 1 above)
Comment (UplinkCoder) [inline std/range/primitives.d]: ... you are right
I had a brain-fart :)

**Comment by UplinkCoder** [comment]
@andralex I do not diverge from the spec anymore.

File: std/range/primitives.d [hunk 2]
```diff
@@ -2104,28 +2104,23 @@ if (isNarrowString!(C[]))
...
-             // 6 and subsquently msbs is 1, which is exactly what we wanted
-             // (skip the invalid byte).
-             uint msbs = 7u - bsr(~c | 1u);
-             if (msbs == 7)
-             {
-                 // Invalid UTF-8, c is 0xFE or 0xFF.
-                 msbs = 1;
-             }
-             str = str.ptr[min(msbs, str.length) .. str.length];
+            str = str.ptr[min(str.length, charWidthTab[c - 192]) .. str.length];
```
Comment (andralex) [inline std/range/primitives.d]: Use `charWidthTab.ptr` here, you've already bounds-checked it.

**Comment by andralex** [comment]
OK, I approve this. Regarding the scrutiny: until we have benchmarking for phobos we need to rely on self-reported performance measurements.

**Comment by UplinkCoder** [comment]
Hmm I cannot merge myself ....

**Comment by wilzbach** [comment]
> until we have benchmarking for phobos we need to rely on self-reported performance measurements.

Yes, but at least we can put the performance measurements into Phobos (or if wanted another common place), s.t. we can improve upon them & others can run the tests themselves.
I bet this PR will be hard to find for future D enthusiasts and the pastelink might even be gone / expired...

**Comment by andralex** [comment]
@wilzbach well, `"until we have benchmarking for phobos".equal("put the performance measurements into Phobos")`

That's a project in and of itself, with its own build target, reporting, regression checks etc. etc. etc. I might be able to put someone on it, but it's way out of the scope of this PR.

**Comment by andralex** [comment]
@UplinkCoder I've invited you

**Comment by UplinkCoder** [comment]
@andralex great merged.

**Comment by wilzbach** [comment]
> @andralex great merged.

You know - autotester is there for a reason :/

**Comment by andralex** [comment]
@UplinkCoder Please create a user id in the autotester and use its "Toggle automerge" feature to pull PRs. Thanks!

**Comment by MartinNowak** [comment]
> Maybe we can use this incident to start benchmarking directly in Phobos. A fancy way would be to run benchmarks for every PR to prevent performance regression, but a simpler approach is to at least collect them at a common place (e.g. benchmarks), s.t. (1) in the future enabling automated testing is possible and (2) other people can use these benchmarks for improving upon them (or manually checking regressions).

We have something like that setup in druntime. A small benchmarks folder and a runner for microbenchmarks.

**Comment by MartinNowak** [comment]
@UplinkCoder please read and/or enhance https://github.com/dlang/phobos/blob/master/CONTRIBUTING.md and the linked wiki pages.

**Comment by schveiguy** [comment]
@UplinkCoder To clarify, in order to use autotester for merging (I'm not sure if this is documented anywhere, it wasn't when I learned how to do it), click on the link to the auto tester in the list of running tests for your PR, e.g.: https://auto-tester.puremagic.com/pull-history.ghtml?projectid=1&repoid=3&pullid=4849

On the upper left corner is a "Login" link. Click on this, and it will link your github account into the auto-tester (not sure if any setup has to happen first time, I don't remember, it just works for me now). Now, the auto tester will provide at the top summary an "Auto-merge" toggle. Click that, and the auto tester will wait for all tests to pass, and merge on your behalf. Please use this when merging any PRs, it (almost) guarantees the master branch will always build.

> please read and/or enhance ...

@MartinNowak I couldn't find anything on auto-tester usage there.

**Pull request merged**